### PR TITLE
Update decision-type logic. Handle new material grants

### PR DESCRIPTION
**Pull request by joofsh**
- Update decision type logic to ensure we are excluding any
     non-new-material allowed issues

connects #1421 

Test Plans:
- [x] Updated tests, ensure granular testing of the decision type methods
- [x] Run `amc_full_grants` pointing at UAT. Verify it runs properly

### CODE REVIEW

File: app/models/appeal.rb [hunk 1]
```diff
@@ -187,15 +187,15 @@ def fetch_documents!(save:)
   end
 
   def partial_grant?
-    status == "Remand" && disposition == "Allowed"
+    status == "Remand" && issues.any?(&:non_new_material_allowed?)
```
Comment (joofsh) [inline app/models/appeal.rb]: this is the crux of the fix. we now specifically check for any issues that are non-new-material grants

File: app/models/vacols/issue.rb [hunk 2]
```diff
@@ -1,19 +1,33 @@
...
 
   # :nocov:
-  def self.format(issue)
+  def self.new_material?(issue)
+    issue["issprog"] == "02" &&
+      issue["isscode"] == "15" &&
+      issue["isslev1"] == "04"
+  end
+
+  def self.load_from_vacols(issue)
```
Comment (joofsh) [inline app/models/vacols/issue.rb]: This part is a little subjective, and I'm open to others' thoughts.

We already have the complex SQL logic in this model for grabbing issues & their descriptions (joining on ISSREF and VFTYPES). Originally, I wanted to isolate the `decision_type` logic from this description stuff. If we are looking at an appeal and only want to determine its decision type, then we don't need the extra two joins. However in practice, these are both currently used when loading the dispatch flow. So we'd be loading issues first to determine decision types, and then a 2nd query for issues with the joins to determine descriptions. So I've gone ahead and aimed to use the same issues SQL query in all cases where we reference issues (that is, only 1 issues VACOLS query). This will be a bit slower, so we should aggressively cache this value (as we already do in Dispatch)

TLDR: For the sake of simplicity we should only have 1  issues query (the complex one we already use for descriptions). It will be a little slower in some places, but I think it's worth it.

File: lib/generators/appeal.rb [hunk 3]
```diff
@@ -92,13 +102,18 @@ def build(attrs = {})
...
       cast_datetime_fields(attrs)
       inaccessible = attrs.delete(:inaccessible)
 
       appeal = Appeal.new(attrs)
 
       vacols_record[:vbms_id] = appeal.vbms_id
 
+      issues_from_template = vacols_record.delete(:issues)
+      set_vacols_issues(appeal: appeal,
+                        issues: issues || issues_from_template)
```
Comment (joofsh) [inline lib/generators/appeal.rb]: This gives us two ways to pass issues. You can directly pass it when you create an appeal, or you can let the issues template load it for you. I think this works pretty well, but open to other ideas

File: app/models/issue.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class Issue
+  include ActiveModel::Model
+
+  attr_accessor :program, :description, :disposition, :new_material
+
+  def non_new_material_allowed?
```
Comment (shanear) [inline app/models/issue.rb]: it's pretty hard to wrap my head around what this actually means. Maybe add a comment to explain it:

```
# Has BVA granted a claim to reopen an issue due to new and material evidence?
# If so, there's no monetary grant associated and thus no EP to create.
```

Also the negation in the method name makes it harder to comprehend. Maybe change the method name to `monetary_allowance?` like so:

```
def monetary_allowance?
  !new_and_material_allowance?
end

private

def new_and_material_allowance?
  ...
end
```

File: app/models/vacols/case.rb [hunk 5]
```diff
@@ -103,11 +103,21 @@ class InvalidLocationError < StandardError; end
...
   VALID_UPDATE_LOCATIONS = %w(50 51 53 54 98).freeze
 
-  JOIN_ISSUE_CNT_REMAND = "
+  JOIN_ISSUE_CNT_ALLOWED = "
     inner join
     (
       select ISSKEY,
-      count(case when ISSDC = '3' then 1 end) ISSUE_CNT_REMAND
+
+      count(case when
```
Comment (shanear) [inline app/models/vacols/case.rb]: this is so damn cryptic, lol

I don't have a good answer to it. I'm just not going to be much help in deciding whether this is right or not.

File: app/models/vacols/issue.rb [hunk 6]
```diff
@@ -1,19 +1,29 @@
+require_relative "../issue"
```
Comment (shanear) [inline app/models/vacols/issue.rb]: the Rails auto loader should handle this.

File: app/models/vacols/issue.rb [hunk 7]
```diff
@@ -1,19 +1,29 @@
...
+
 class VACOLS::Issue < VACOLS::Record
   self.table_name = "vacols.issues"
   self.sequence_name = "vacols.issseq"
   self.primary_key = "isskey"
 
   # :nocov:
-  def self.format(issue)
+  def self.new_material?(issue)
```
Comment (shanear) [inline app/models/vacols/issue.rb]: Do we know the significance of these numbers? It kind of makes me cringe to see this logic here with a `#nocov`. It be nice if this sort of logic was more semantic and inside `Issue`instead.

File: app/models/vacols/issue.rb (hunk 2 above)
Comment (shanear) [inline app/models/vacols/issue.rb]: I like this approach. Way more flexible & easy to reason about too. 👍 👍 👍

File: app/models/vacols/issue.rb [hunk 8]
```diff
@@ -1,19 +1,29 @@
...
+
+  def self.load_from_vacols(issue)
     description = ["#{issue['isscode']} - #{issue['isscode_label']}"]
     description.push("#{issue['isslev1']} - #{issue['isslev1_label']}") if issue["isslev1"]
     description.push("#{issue['isslev2']} - #{issue['isslev2_label']}") if issue["isslev2"]
     description.push("#{issue['isslev3']} - #{issue['isslev3_label']}") if issue["isslev3"]
-    {
+
+    Issue.new(
       program: "#{issue['issprog']} - #{issue['issprog_label']}",
```
Comment (shanear) [inline app/models/vacols/issue.rb]: can we move this logic to `Issue`. It'd be nice if the values we pass to `Issue.new` are as "raw" as possible so we can test stuff in `Issue`.

File: spec/models/appeal_spec.rb [hunk 9]
```diff
@@ -251,58 +251,103 @@
...
+      it { is_expected.to be_falsey }
+    end
+
+    context "is false if allowed issues are new material" do
+      let(:issues) { [{ new_material: true, disposition: "Allowed" }] }
+
       it { is_expected.to be_falsey }
     end
 
     context "is true" do
```
Comment (shanear) [inline spec/models/appeal_spec.rb]: this is a strange `context`

File: spec/models/appeal_spec.rb [hunk 10]
```diff
@@ -251,58 +251,103 @@
...
-    context "is false" do
-      let(:appeal) { Appeal.new(vacols_id: "123", status: "Complete", disposition: "Allowed") }
+
+    context "is false if no allowed issues" do
+      let(:issues) { [{ new_material: false, disposition: "Remand" }] }
+
+      it { is_expected.to be_falsey }
+    end
+
+    context "is false if allowed issues are new material" do
```
Comment (shanear) [inline spec/models/appeal_spec.rb]: dont need "is false" here. Rspec will parse line 266 and put that in the output. As a rule of thumb, `context` statements should mostly start with "when", because they are defining a condition not an assertion

`context "when allowed issues are new material"`

File: spec/models/appeal_spec.rb [hunk 11]
```diff
@@ -251,58 +251,103 @@
...
+        Generators::Appeal.build(vacols_id: "123", status: "Complete", issues: issues)
+      end
+      it { is_expected.to be_falsey }
+    end
+
+    context "is true if non-new-material allowed issues" do
+      let(:issues) { [{ new_material: false, disposition: "Allowed" }] }
+      let(:appeal) do
+        Generators::Appeal.build(vacols_id: "123", status: "Complete", issues: issues)
+      end
```
Comment (shanear) [inline spec/models/appeal_spec.rb]: could use contexts here to wrap similar `let` statements in this test.

i.e.
```
let(:appeal) do
  Generators::Appeal.build(vacols_id: "123", status: status, issues: issues)
end

let(:issues) { [{ new_material: false, disposition: "Allowed" }] }

context "when status is not Complete" do
  status { "Remand" }

  ...
end

context "when status is Complete" do
  status { "Complete" }

  context "when all issues are new-material allowed" do
    let(:issues) { [{ new_material: true, disposition: "Allowed" }] }
    ...
  end

  context "when at least one issue is not new-material allowed" do
    ...
  end
end
```

File: lib/fakes/appeal_repository.rb [hunk 12]
```diff
@@ -160,22 +160,8 @@ def self.uncertify(_appeal)
...
-        "isslev1_label" => "All Others",
-        "isslev2" => "30",
-        "isslev2_label" => "Post-traumatic stress disorder",
-        "isslev3" => nil,
-        "isslev3_label" => nil,
-        "issdc" => "Allowed"
-      )
-    ]
+  def self.issues(vacols_id)
+    (issue_records || {})[vacols_id] || []
```
Comment (joofsh) [inline lib/fakes/appeal_repository.rb]: getting this meme from you made this entire PR worth it :)

File: app/models/vacols/case.rb (hunk 5 above)
Comment (joofsh) [inline app/models/vacols/case.rb]: I'm 99% sure it's right. I tested it on UAT :)

File: app/models/issue.rb (hunk 4 above)
Comment (joofsh) [inline app/models/issue.rb]: I'm happy to add comments, but I think what you suggest is more confusing. We don't mention the word `monetary` anywhere in our code yet.  "New Material" (and "Non new material") are the exact terms used internally by attorneys/judges. I'd rather stay as close to the real language as possible

Thoughts?

File: app/models/vacols/issue.rb (hunk 6 above)
Comment (joofsh) [inline app/models/vacols/issue.rb]: I thought so too, but it doesn't:
```
LoadError: Unable to autoload constant Issue, expected /Users/jd/src/caseflow-certification/app/models/vacols/issue.rb to define it
```

File: app/models/vacols/issue.rb (hunk 7 above)
Comment (joofsh) [inline app/models/vacols/issue.rb]: these numbers, together, signify a new material issue.  

I thought you specifically wanted to "mask" vacols column names from our models, as we do with the appeal model. I could directly pass all these values to our Issue model, but then we'd be directly referencing `isslev1` and other "random" column names in our code.

File: app/models/issue.rb (hunk 4 above)
Comment (shanear) [inline app/models/issue.rb]: ohh okay, I'm cool with that, maybe write this in the comment too:

>"New Material" (and "Non new material") are the exact terms used internally by attorneys/judges.

It also might be clearer to have both:
```
def non_new_material?
  !new_material?
end

def new_material?
   ...
end
```

File: app/models/vacols/case.rb (hunk 5 above)
Comment (shanear) [inline app/models/vacols/case.rb]: it's not as much the sql as it is the `ISSPROG = '02'` that i have no idea what it means

File: app/models/vacols/issue.rb (hunk 6 above)
Comment (shanear) [inline app/models/vacols/issue.rb]: Try using `::Issue` instead of `Issue`?

http://stackoverflow.com/questions/32000559/ruby-on-rails-multiple-classes-with-same-name-under-different-namespace

File: app/models/vacols/case.rb [hunk 13]
```diff
@@ -152,7 +161,7 @@ def self.remands_ready_for_claims_establishment
   end
 
   def self.amc_full_grants(outcoded_after:)
-    VACOLS::Case.joins(:folder, :correspondent, JOIN_ISSUE_CNT_REMAND)
+    VACOLS::Case.joins(:folder, :correspondent, JOIN_ISSUE_CNT_ALLOWED)
```
Comment (cmgiven) [inline app/models/vacols/case.rb]: Can we continue to check that there are no remands present, in addition to checking that there is at least one non-n&m allowance?

File: app/models/vacols/issue.rb (hunk 6 above)
Comment (joofsh) [inline app/models/vacols/issue.rb]: It looks like rails doesn't like 2 models with the same file name. Both `VACOLS::Issue` and `Issue` were in a file called `issue.rb`. Changing the name of one of the files resolved the issue.  

Does that sound right to you?

File: app/models/issue.rb [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity
+    def load_from_vacols(hash)
+      description = ["#{hash["isscode"]} - #{hash["isscode_label"]}"]
+      description.push("#{hash["isslev1"]} - #{hash["isslev1_label"]}") if hash["isslev1"]
+      description.push("#{hash["isslev2"]} - #{hash["isslev2_label"]}") if hash["isslev2"]
+      description.push("#{hash["isslev3"]} - #{hash["isslev3_label"]}") if hash["isslev3"]
+
+      category_code = hash["isslev1"] || hash["isslev2"] || hash["isslev3"]
+
+      disposition = (VACOLS::Case::DISPOSITIONS[hash["issdc"]] || "other")
```
Comment (shanear) [inline app/models/issue.rb]: I noticed your converting this to a symbol, I like that, but it seems like the original hash should just be a symbol.

Only issue with that, is that we are using the disposition string of appeals directly in the Form 8 UI, so we'd have to convert it to a string at some point. That conversion feels better to me.

Comment (shanear) [inline app/models/issue.rb]: Just a thought to chew on, def not saying we should do this now. Maybe create a "strings to symbols" tech improvement ticket?

File: app/models/issue.rb [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    !new_material?
+  end
+
+  class << self
+    # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity
+    def load_from_vacols(hash)
+      description = ["#{hash["isscode"]} - #{hash["isscode_label"]}"]
+      description.push("#{hash["isslev1"]} - #{hash["isslev1_label"]}") if hash["isslev1"]
+      description.push("#{hash["isslev2"]} - #{hash["isslev2_label"]}") if hash["isslev2"]
+      description.push("#{hash["isslev3"]} - #{hash["isslev3_label"]}") if hash["isslev3"]
```
Comment (shanear) [inline app/models/issue.rb]: we could get rid of those `rubocop:disable`s by moving all of this to a helper private function

File: client/test/karma/util/StringUtil-test.js [hunk 16]
```diff
@@ -19,4 +19,18 @@ describe('StringUtil', () => {
       expect(StringUtil.leftPad('12', 4, '0')).to.eq('0012');
     });
   });
+
+  context('.titleCase', () => {
```
Comment (shanear) [inline client/test/karma/util/StringUtil-test.js]: beautiful test :1st_place_medal:

File: lib/generators/appeal.rb [hunk 17]
```diff
@@ -49,19 +49,29 @@ def vacols_record_templates
...
+            { disposition: :remanded, program: :compensation, type: :service_connection, category: :knee }
+          ]
         },
         partial_grant_decided: {
           status: "Remand",
           disposition: "Allowed",
-          decision_date: 7.days.ago
+          decision_date: 7.days.ago,
+          issues: [
+            { disposition: :remanded, program: :compensation, type: :service_connection, category: :knee },
```
Comment (shanear) [inline lib/generators/appeal.rb]: should we have a template for this?
